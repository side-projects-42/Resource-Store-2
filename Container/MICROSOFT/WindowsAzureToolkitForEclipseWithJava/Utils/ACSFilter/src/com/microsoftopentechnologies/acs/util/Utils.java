/**
 * Copyright (c) Microsoft Corporation
 * 
 * All rights reserved. 
 * 
 * MIT License
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files 
 * (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, 
 * publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, 
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR 
 * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH 
 * THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.microsoftopentechnologies.acs.util; 

import java.io.OutputStream;
import java.io.StringWriter;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.security.spec.AlgorithmParameterSpec;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.TimeZone;
import java.util.logging.Level;
import java.util.logging.Logger;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;

import org.w3c.dom.Element;
import org.w3c.dom.Node;

public class Utils {
	private static final Logger LOG = Logger.getLogger(Utils.class.getName());
	public static final String UTF8_ENCODING = "UTF-8";
	private static final String ACS_TIME_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
	private static final SimpleDateFormat GMT_DATE_FMT = new SimpleDateFormat(ACS_TIME_FORMAT);
	
	public static final void logError(String text, Throwable error, Logger logger) {
		if (logger != null && logger.isLoggable(Level.SEVERE)) {
			logger.log(Level.SEVERE, text);
		}
	}

	public static final void logDebug(String text, Logger logger) {
		if (logger != null && logger.isLoggable(Level.FINE)) {
			logger.log(Level.FINE, text);
		}
	}

	public static final void logInfo(String text, Logger logger) {
		if (logger != null && logger.isLoggable(Level.INFO)) {
			logger.log(Level.INFO, text);
		}
	}

	public static final void logWarn(String text, Throwable error, Logger logger)	{
		if (logger != null && logger.isLoggable(Level.WARNING)) {
			logger.log(Level.WARNING, text);
		}
	}

	static {
		// Use GMT time zone, instead of the local time zone
		GMT_DATE_FMT.setTimeZone(TimeZone.getTimeZone("GMT"));
	}

	/*
	 * Just to move the try-catch to a separate method and keep the code clean at
	 * other places where URL decoding is required.
	 */
	public static final String urlDecode(String encodedString)	{
		String decodedString = null;
		try {
			decodedString =  URLDecoder.decode(encodedString, UTF8_ENCODING);
		} catch (UnsupportedEncodingException e) {
			//Ignore it.. Never arise...
			logError("UTF-8 charset not found on the JRE.", e, LOG);
		}
		return decodedString;
	}

	public static final String urlEncode(String input) {
		String encString = null;
		try {
			encString =  URLEncoder.encode(input, UTF8_ENCODING);
		} catch (UnsupportedEncodingException e) {
			//Ignore it.. Never arise...
			logError("UTF-8 charset not found on the JRE.", e, LOG);
		}
		return encString;
	}

	public static final byte[] getUTF8Bytes(String input) {
		byte[] utf8Bytes = null;
		try {
			utf8Bytes = input.getBytes(UTF8_ENCODING);
		} catch (UnsupportedEncodingException e) {
			// Ignore it.. Never arise...
			logError("UTF-8 charset not found on the JRE.", e, LOG);
		}
		return utf8Bytes;
	}

	public static final String getStringFromUTF8Bytes(byte[] input)	{
		String utf8Msg = null;
		try {
			utf8Msg =  new String(input, UTF8_ENCODING);
		} catch (UnsupportedEncodingException e) {
			logError("UTF-8 charset not found on the JRE.", e, LOG);
		}
		return utf8Msg;
	}

	// Synchronized as SimpleDateFormat is not thread safe..shall I create an instance on demand?
	public synchronized static Date parseTimeFromXSDTIMEString(String timeAsString) throws ParseException {
		return GMT_DATE_FMT.parse(timeAsString);
	}

	// Using Transformer.. Just in case.. we need to log XML element
	public static void writeXMLToStream(Node node, OutputStream outputStream, boolean indented) throws TransformerException	{
		TransformerFactory transFact = TransformerFactory.newInstance();
		Transformer transformer = transFact.newTransformer();
		String indent = indented ? "yes" : "no";
		transformer.setOutputProperty(OutputKeys.INDENT, indent);
		transformer.transform(new DOMSource(node), new StreamResult(outputStream));
	}
	

	public static String getXMLStringFromNode(Element node) throws Exception {
		DOMSource domSource = new DOMSource(node);
		
		StringWriter writer = new StringWriter();
		StreamResult result = new StreamResult(writer);
		TransformerFactory tf = TransformerFactory.newInstance();
		Transformer transformer = tf.newTransformer();
		transformer.transform(domSource, result);
		return result.getWriter().toString();
	}
	
	/*
	 * If the inputText size is larger than the size limit, this method splits the input text into blocks of size equal to the
	 * limit. The last part may be smaller than the size limit. Value of 0 for size limit is not expected, it produces an exception.
	 */
	public static String[] splitText(String inputText, int blockSizeLimit)	{
		int inputTextSize = inputText.length();
		// Dividing two times.. is there an easier way to find both the remaider and quotient?
		int number_parts = ((inputTextSize % blockSizeLimit) == 0) ? (inputTextSize / blockSizeLimit) : ((inputTextSize / blockSizeLimit) + 1);
		String[] splitParts = new String[number_parts];
		for (int i = 0; i < (number_parts - 1); i++) {
			int beginIndex = blockSizeLimit * i;
			int endIndex = beginIndex + blockSizeLimit;
			splitParts[i] = inputText.substring(beginIndex, endIndex);
		}
		splitParts[number_parts - 1] = inputText.substring(blockSizeLimit * (number_parts - 1));
		return splitParts;
	}

	public static String removeCRLFsInBase64EncodedText(String base64EncodedText)	{
		return base64EncodedText.replaceAll("\\s", ""); // Replace all white spaces
	}
	
	public static SecretKey getSecretKey(String key) {
		byte[] keyInBytes = Base64.decode(key);;
		SecretKey secretKey = new SecretKeySpec(keyInBytes, "AES");
		return secretKey;
	}
	
	public static byte[] encrypt(SecretKey secretKey,byte[] assertionContent) throws Exception {
		byte[] initializationVector = new byte[] {0, 9, 8, 7, 6, 2, 3, 4, 5, 0,1, 5, 7, 8, 9, 4};
		AlgorithmParameterSpec ivParameterSpec = new IvParameterSpec(initializationVector);
		byte[] encryptedText = null;
		
		try {		
			Cipher encrypter = Cipher.getInstance("AES/CBC/PKCS5Padding");
			encrypter.init(Cipher.ENCRYPT_MODE, secretKey, ivParameterSpec);
			encryptedText = encrypter.doFinal(assertionContent);			
		}catch(Exception e) {
			//There is possibility of NoSuchPaddingException , NoSuchAlgorithmException , InvalidKeyException , InvalidAlgorithmParameterException,
			//IllegalBlockSizeException , BadPaddingException. For this method any such exception is considered has failure and more over not taking 
			//specific action on type of exception , hence catching all in one block
			logError("NoSuchPaddingException", e, LOG);
			throw e;
		}
		return encryptedText;		
	}
	
	public static byte[] decrypt(SecretKey secretKey,byte[] encryptedText) throws Exception {
		byte[] decryptedText = null;
		byte[] initializationVector = new byte[] {0, 9, 8, 7, 6, 2, 3, 4, 5, 0,1, 5, 7, 8, 9, 4};
		AlgorithmParameterSpec ivParameterSpec = new IvParameterSpec(initializationVector);
		
		try {		
			Cipher decrypter = Cipher.getInstance("AES/CBC/PKCS5Padding");
			decrypter.init(Cipher.DECRYPT_MODE, secretKey, ivParameterSpec);
			decryptedText = decrypter.doFinal(encryptedText);				
		}catch(Exception e) {
			//There is possibility of NoSuchPaddingException , NoSuchAlgorithmException , InvalidKeyException , InvalidAlgorithmParameterException,
			//IllegalBlockSizeException , BadPaddingException. For this method any such exception is considered has failure and more over not taking 
			//specific action on type of exception , hence catching all in one block
			logError("NoSuchPaddingException", e, LOG);
			throw e;
		}
		return decryptedText;
		
	}

	

}
