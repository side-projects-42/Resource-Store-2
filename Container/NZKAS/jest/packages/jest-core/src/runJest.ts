/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import chalk from 'chalk';
import {sync as realpath} from 'realpath-native';
import {CustomConsole} from '@jest/console';
import {formatTestResults} from 'jest-util';
import exit from 'exit';
import fs from 'graceful-fs';
import {JestHook, JestHookEmitter} from 'jest-watcher';
import {Context} from 'jest-runtime';
import {Test} from 'jest-runner';
import {Config} from '@jest/types';
import {
  AggregatedResult,
  makeEmptyAggregatedTestResult,
} from '@jest/test-result';
import {ChangedFiles, ChangedFilesPromise} from 'jest-changed-files';
import getNoTestsFoundMessage from './getNoTestsFoundMessage';
import runGlobalHook from './runGlobalHook';
import SearchSource from './SearchSource';
import TestScheduler, {TestSchedulerContext} from './TestScheduler';
import TestSequencer from './TestSequencer';
import FailedTestsCache from './FailedTestsCache';
import collectNodeHandles from './collectHandles';
import TestWatcher from './TestWatcher';
import {TestRunData, Filter} from './types';

const getTestPaths = async (
  globalConfig: Config.GlobalConfig,
  context: Context,
  outputStream: NodeJS.WritableStream,
  changedFiles: ChangedFiles | undefined,
  jestHooks: JestHookEmitter,
  filter?: Filter,
) => {
  const source = new SearchSource(context);
  const data = await source.getTestPaths(globalConfig, changedFiles, filter);

  if (!data.tests.length && globalConfig.onlyChanged && data.noSCM) {
    new CustomConsole(outputStream, outputStream).log(
      'Jest can only find uncommitted changed files in a git or hg ' +
        'repository. If you make your project a git or hg ' +
        'repository (`git init` or `hg init`), Jest will be able ' +
        'to only run tests related to files changed since the last ' +
        'commit.',
    );
  }

  const shouldTestArray = await Promise.all(
    data.tests.map(test =>
      jestHooks.shouldRunTestSuite({
        config: test.context.config,
        duration: test.duration,
        testPath: test.path,
      }),
    ),
  );

  const filteredTests = data.tests.filter((_test, i) => shouldTestArray[i]);

  return {...data, allTests: filteredTests.length, tests: filteredTests};
};

type ProcessResultOptions = Pick<
  Config.GlobalConfig,
  'json' | 'outputFile' | 'testResultsProcessor'
> & {
  collectHandles?: () => Array<Error>;
  onComplete?: (result: AggregatedResult) => void;
  outputStream: NodeJS.WritableStream;
};

const processResults = (
  runResults: AggregatedResult,
  options: ProcessResultOptions,
) => {
  const {
    outputFile,
    json: isJSON,
    onComplete,
    outputStream,
    testResultsProcessor,
    collectHandles,
  } = options;

  if (collectHandles) {
    runResults.openHandles = collectHandles();
  } else {
    runResults.openHandles = [];
  }

  if (testResultsProcessor) {
    runResults = require(testResultsProcessor)(runResults);
  }
  if (isJSON) {
    if (outputFile) {
      const cwd = realpath(process.cwd());
      const filePath = path.resolve(cwd, outputFile);

      fs.writeFileSync(filePath, JSON.stringify(formatTestResults(runResults)));
      outputStream.write(
        `Test results written to: ${path.relative(cwd, filePath)}\n`,
      );
    } else {
      process.stdout.write(JSON.stringify(formatTestResults(runResults)));
    }
  }

  return onComplete && onComplete(runResults);
};

const testSchedulerContext: TestSchedulerContext = {
  firstRun: true,
  previousSuccess: true,
};

export default (async function runJest({
  contexts,
  globalConfig,
  outputStream,
  testWatcher,
  jestHooks = new JestHook().getEmitter(),
  startRun,
  changedFilesPromise,
  onComplete,
  failedTestsCache,
  filter,
}: {
  globalConfig: Config.GlobalConfig;
  contexts: Array<Context>;
  outputStream: NodeJS.WritableStream;
  testWatcher: TestWatcher;
  jestHooks?: JestHookEmitter;
  startRun: (globalConfig: Config.GlobalConfig) => void;
  changedFilesPromise?: ChangedFilesPromise;
  onComplete: (testResults: AggregatedResult) => void;
  failedTestsCache?: FailedTestsCache;
  filter?: Filter;
}) {
  const sequencer = new TestSequencer();
  let allTests: Array<Test> = [];

  if (changedFilesPromise && globalConfig.watch) {
    const {repos} = await changedFilesPromise;

    const noSCM = (Object.keys(repos) as Array<
      keyof ChangedFiles['repos']
    >).every(scm => repos[scm].size === 0);
    if (noSCM) {
      process.stderr.write(
        '\n' +
          chalk.bold('--watch') +
          ' is not supported without git/hg, please use --watchAll ' +
          '\n',
      );
      exit(1);
    }
  }

  const testRunData: TestRunData = await Promise.all(
    contexts.map(async context => {
      const matches = await getTestPaths(
        globalConfig,
        context,
        outputStream,
        changedFilesPromise && (await changedFilesPromise),
        jestHooks,
        filter,
      );
      allTests = allTests.concat(matches.tests);

      return {context, matches};
    }),
  );

  allTests = sequencer.sort(allTests);

  if (globalConfig.listTests) {
    const testsPaths = Array.from(new Set(allTests.map(test => test.path)));
    if (globalConfig.json) {
      console.log(JSON.stringify(testsPaths));
    } else {
      console.log(testsPaths.join('\n'));
    }

    onComplete && onComplete(makeEmptyAggregatedTestResult());
    return null;
  }

  if (globalConfig.onlyFailures && failedTestsCache) {
    allTests = failedTestsCache.filterTests(allTests);
    globalConfig = failedTestsCache.updateConfig(globalConfig);
  }

  const hasTests = allTests.length > 0;

  if (!hasTests) {
    const noTestsFoundMessage = getNoTestsFoundMessage(
      testRunData,
      globalConfig,
    );

    if (
      globalConfig.passWithNoTests ||
      globalConfig.findRelatedTests ||
      globalConfig.lastCommit ||
      globalConfig.onlyChanged
    ) {
      new CustomConsole(outputStream, outputStream).log(noTestsFoundMessage);
    } else {
      new CustomConsole(outputStream, outputStream).error(noTestsFoundMessage);

      exit(1);
    }
  } else if (
    allTests.length === 1 &&
    globalConfig.silent !== true &&
    globalConfig.verbose !== false
  ) {
    const newConfig: Config.GlobalConfig = {...globalConfig, verbose: true};
    globalConfig = Object.freeze(newConfig);
  }

  let collectHandles;

  if (globalConfig.detectOpenHandles) {
    collectHandles = collectNodeHandles();
  }

  if (hasTests) {
    await runGlobalHook({allTests, globalConfig, moduleName: 'globalSetup'});
  }

  if (changedFilesPromise) {
    testSchedulerContext.changedFiles = (await changedFilesPromise).changedFiles;
  }

  const results = await new TestScheduler(
    globalConfig,
    {startRun},
    testSchedulerContext,
  ).scheduleTests(allTests, testWatcher);

  sequencer.cacheResults(allTests, results);

  if (hasTests) {
    await runGlobalHook({allTests, globalConfig, moduleName: 'globalTeardown'});
  }

  return processResults(results, {
    collectHandles,
    json: globalConfig.json,
    onComplete,
    outputFile: globalConfig.outputFile,
    outputStream,
    testResultsProcessor: globalConfig.testResultsProcessor,
  });
});
