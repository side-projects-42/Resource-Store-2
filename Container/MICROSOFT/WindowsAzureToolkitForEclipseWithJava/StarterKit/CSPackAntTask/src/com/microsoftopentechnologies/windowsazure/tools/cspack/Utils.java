/**
 * Copyright (c) Microsoft Corporation
 * 
 * All rights reserved. 
 * 
 * MIT License
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files 
 * (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, 
 * publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, 
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR 
 * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH 
 * THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */


package com.microsoftopentechnologies.windowsazure.tools.cspack;

import com.github.jknack.handlebars.*;
import com.github.jknack.handlebars.context.JavaBeanValueResolver;
import com.microsoftopentechnologies.windowsazure.tools.build.WindowsAzurePackage;
import org.apache.tools.ant.BuildException;
import org.apache.tools.ant.util.Base64Converter;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Unmarshaller;
import java.io.*;
import java.net.JarURLConnection;
import java.net.URL;
import java.net.URLConnection;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

public class Utils {

    private static String convertByteArrayToHexString(byte[] arrayBytes) {
        StringBuffer stringBuffer = new StringBuffer();
        for (int i = 0; i < arrayBytes.length; i++) {
            stringBuffer.append(Integer.toString((arrayBytes[i] & 0xff) + 0x100, 16)
                    .substring(1));
        }
        return stringBuffer.toString();
    }

    /**
     * Calculate Sha256 hash
     */
    public static String calcHash(File file) throws IOException {
        int bufferSize = 16384;
        String output;
        try {
            BufferedInputStream is = new BufferedInputStream(new FileInputStream(file));
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte [] buffer = new byte[bufferSize];
            int sizeRead = -1;
            while ((sizeRead = is.read(buffer)) != -1) {
                digest.update(buffer, 0, sizeRead);
            }
            is.close();

            byte [] hash = null;
            hash = new byte[digest.getDigestLength()];
            hash = digest.digest();
            output = new Base64Converter().encode(hash);
        } catch (FileNotFoundException e) {
            throw new BuildException(e);
        } catch (NoSuchAlgorithmException e) {
            throw new BuildException(e);
        }
        return output;
    }

    public static String calcHexHash(File file) throws IOException {
        int bufferSize = 16384;
        String output;
        try {
            BufferedInputStream is = new BufferedInputStream(new FileInputStream(file));
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte [] buffer = new byte[bufferSize];
            int sizeRead = -1;
            while ((sizeRead = is.read(buffer)) != -1) {
                digest.update(buffer, 0, sizeRead);
            }
            is.close();

            byte [] hash = null;
            hash = new byte[digest.getDigestLength()];
            hash = digest.digest();
            output = convertByteArrayToHexString(hash).toUpperCase();
        } catch (FileNotFoundException e) {
            throw new BuildException(e);
        } catch (NoSuchAlgorithmException e) {
            throw new BuildException(e);
        }
        return output;
    }

    /**
     * Zip directory (source can't be a file)
     */
    public static void zipDirectory(File sourceDir, File zipFile, WindowsAzurePackage waPackage) {
        for (File file : sourceDir.listFiles()) {
            waPackage.zipFile(file, zipFile);
            delete(file);
        }
    }

    /**
     * Compresses the given directory and all its sub-directories into a ZIP file.
     * <p>
     * The ZIP file must not be a directory and its parent directory must exist.
     *
     * @param sourceDir
     *          root directory.
     * @param targetZipName
     *          name of ZIP file that will be created or overwritten.
     */
    public static void pack(File sourceDir, String targetZipName, boolean deleteSource) {
        if (!sourceDir.exists()) {
            throw new RuntimeException("Given file '" + sourceDir + "' doesn't exist!");
        }
        ZipOutputStream out = null;
        try {
            String[] filenames = sourceDir.list();
            if (filenames == null) {
                if (!sourceDir.exists()) {
                    throw new RuntimeException("Given file '" + sourceDir + "' doesn't exist!");
                }
                throw new IOException("Given file is not a directory '" + sourceDir + "'");
            }
            out = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(new File(targetZipName))));
            pack(filenames, sourceDir, out, "");
            if (deleteSource) {
                for (String filename : filenames) {
                    delete(new File(sourceDir, filename));
                }
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        } finally {
            closeQuietly(out);
        }
    }

    /**
     * Compresses the given directory and all its sub-directories into a ZIP file.
     *
     * @param filenames
     *          files and directories names.
     * @param dir
     *          root directory.
     * @param out
     *          ZIP output stream.
     * @param pathPrefix
     *          prefix to be used for the entries.
     */
    private static void pack(String[] filenames, File dir, ZipOutputStream out, String pathPrefix) throws IOException {
        for (int i = 0; i < filenames.length; i++) {
            String filename = filenames[i];
            File file = new File(dir, filename);
            boolean isDir = file.isDirectory();
            String path = pathPrefix + file.getName();
            if (isDir) {
                path += "/";
            }
            // Create a ZIP entry
            ZipEntry zipEntry = new ZipEntry(path);
            if (!isDir) {
                zipEntry.setSize(file.length());
                zipEntry.setTime(file.lastModified());
            }
            out.putNextEntry(zipEntry);
            // Copy the file content
            if (!isDir) {
                copy(file, out);
            }
            out.closeEntry();
            // Traverse the directory
            if (isDir) {
                String[] dirFilenames = file.list();
                if (dirFilenames == null) {
                    if (!dir.exists()) {
                        throw new RuntimeException("Given file '" + dir + "' doesn't exist!");
                    }
                    throw new IOException("Given file is not a directory '" + dir + "'");
                }
                pack(dirFilenames, file, out, path);
            }
        }
    }

    /**
     * Copies the given file into an output stream.
     *
     * @param file input file (must exist).
     * @param out output stream.
     */
    public static void copy(File file, OutputStream out) throws IOException {
        FileInputStream in = new FileInputStream(file);
        try {
            byte[] buffer = new byte[4*1024];
            long count = 0;
            int n = 0;
            InputStream input = new BufferedInputStream(in);
            while (-1 != (n = input.read(buffer))) {
                out.write(buffer, 0, n);
                count += n;
            }
        }
        finally {
            closeQuietly(in);
        }
    }

    /**
     * Unconditionally close an <code>InputStream</code>.
     * @param stream the InputStream to close, may be null or already closed
     */
    public static void closeQuietly(Closeable stream) {
        try {
            if (stream != null) {
                stream.close();
            }
        } catch (IOException ioe) {
            // ignore
        }
    }

    public static void delete(File f) {
        if (f.isDirectory()) {
            for (File c : f.listFiles())
                delete(c);
        }
        if (!f.delete())
            throw new RuntimeException("Failed to delete file: " + f);
    }

    public static void generateFileList(Collection<File> fileList, File node) {
        //add file only
        if (node.isFile()) {
            fileList.add(node.getAbsoluteFile());
        }

        if (node.isDirectory()) {
            String[] subNote = node.list();
            for (String filename : subNote) {
                generateFileList(fileList, new File(node, filename));
            }
        }
    }

    public static String generateUID(String name) {
        UUID uuid = UUID.nameUUIDFromBytes(name.getBytes());
        return uuid.toString().replaceAll("-", "");
    }

    static void createDirectory(String name) {
        File serviceDefinitionDirectory = new File(name);
        if (!serviceDefinitionDirectory.exists()) {
            serviceDefinitionDirectory.mkdir();
        }
    }

    public static void applyTemplateWithPath(String templateName, Object model, String deployPath) throws IOException {
        applyTemplate(templateName, model, deployPath + File.separator + templateName);
    }

    public static void applyTemplate(String templateName, Object model, String fileName) throws IOException {
//        System.out.println("templateName=" + templateName + "; fileName=" + fileName);
        BufferedWriter fileWriter = new BufferedWriter(new FileWriter(fileName));
        try {
            Handlebars handlebars = new Handlebars();
            Template template = handlebars.compile("/templates/" + templateName);

            Context context = Context.newBuilder(model).resolver(JavaBeanValueResolver.INSTANCE).build();
            template.apply(context, fileWriter);
//            System.out.println(template.apply(context));
        } finally {
            fileWriter.flush();
            fileWriter.close();
        }
    }

    public static List<String> getJarEntries(String jarName, String entryName) throws IOException {
        List<String> files = new ArrayList<String>();
        System.out.println("entryName = " + entryName);
        JarURLConnection urlConnection = (JarURLConnection) new URL("jar:file:" + jarName + "!/" + entryName).openConnection();
//        URLConnection urlConnection = originUrl.openConnection();
            JarURLConnection jarConnection = ((JarURLConnection)urlConnection);
            JarFile jarFile = jarConnection.getJarFile();
            Enumeration<JarEntry> entries = jarFile.entries();

            while (entries.hasMoreElements()) {
                JarEntry entry = entries.nextElement();
                if (entry.getName().startsWith(jarConnection.getEntryName())) {
                    if (!entry.isDirectory()) {
                        files.add(entry.getName());
                    }
                }
            }
        return files;
    }

    public static void copyJarEntry(String source, File destination) throws IOException {
        InputStream stream = Utils.class.getResourceAsStream(source);
        if (stream == null) {
            throw new BuildException(source + " not found");
        }
        OutputStream resStreamOut = null;
        int readBytes;
        byte[] buffer = new byte[4096];
        try {
            resStreamOut = new FileOutputStream(destination);
            while ((readBytes = stream.read(buffer)) > 0) {
                resStreamOut.write(buffer, 0, readBytes);
            }
        } finally {
            stream.close();
            if (resStreamOut != null) {
                resStreamOut.close();
            }
        }
    }

    public static <T> T parseXmlFile(Class<T> type, String filePath) throws JAXBException {
        T instance = null;
        File file = new File(filePath);
        JAXBContext jaxbContext = JAXBContext.newInstance(type);

        Unmarshaller jaxbUnmarshaller = jaxbContext.createUnmarshaller();
        instance = (T) jaxbUnmarshaller.unmarshal(file);
        return instance;
    }
}
