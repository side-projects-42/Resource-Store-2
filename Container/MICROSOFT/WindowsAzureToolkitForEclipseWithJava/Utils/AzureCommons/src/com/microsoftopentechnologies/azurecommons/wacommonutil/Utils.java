/**
 * Copyright (c) Microsoft Corporation
 * 
 * All rights reserved. 
 * 
 * MIT License
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files 
 * (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, 
 * publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, 
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR 
 * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH 
 * THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.microsoftopentechnologies.azurecommons.wacommonutil;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.text.SimpleDateFormat;
import java.util.Date;

public class Utils {

	public static String replaceLastSubString(String location, String find,
			String replaceWith) {
		if (location == null || location.isEmpty())
			return location;

		int lastIndex = location.lastIndexOf(find);

		if (lastIndex < 0)
			return location;

		String end = location.substring(lastIndex).replaceFirst(find,
				replaceWith);
		return location.substring(0, lastIndex) + end;
	}

	public static String getDefaultCNName() {
		SimpleDateFormat dateFormat = new SimpleDateFormat("yyyyMMddHHmmss");
		Date now = new Date();
		return "Self Signed Certificate " + dateFormat.format(now);
	}

	/**
	 * This method is used for invoking native commands.
	 * 
	 * @param command
	 *            :- command to invoke.
	 * @param ignoreErrorStream
	 *            : Boolean which controls whether to throw exception or not
	 *            based on error stream.
	 * @return result :- depending on the method invocation.
	 * @throws Exception
	 * @throws IOException
	 */
	public static String cmdInvocation(String[] command,
			boolean ignoreErrorStream) throws Exception, IOException {
		String result = "";
		String error = "";
		InputStream inputStream = null;
		InputStream errorStream = null;
		BufferedReader br = null;
		BufferedReader ebr = null;
		try {
			Process process = new ProcessBuilder(command).start();
			;
			inputStream = process.getInputStream();
			errorStream = process.getErrorStream();
			br = new BufferedReader(new InputStreamReader(inputStream));
			result = br.readLine();
			process.waitFor();
			ebr = new BufferedReader(new InputStreamReader(errorStream));
			error = ebr.readLine();
			if (error != null && (!error.equals(""))) {
				// To do - Log error message

				if (!ignoreErrorStream) {
					throw new Exception(error, null);
				}
			}
		} catch (Exception e) {
			throw new Exception("Exception occurred while invoking command", e);
		} finally {
			if (inputStream != null) {
				inputStream.close();
			}
			if (errorStream != null) {
				errorStream.close();
			}
			if (br != null) {
				br.close();
			}
			if (ebr != null) {
				ebr.close();
			}
		}
		return result;
	}

}
