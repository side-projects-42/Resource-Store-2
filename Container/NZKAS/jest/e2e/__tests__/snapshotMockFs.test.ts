/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import rimraf from 'rimraf';
import {wrap} from 'jest-snapshot-serializer-raw';
import {extractSummary} from '../Utils';
import {json as runJestJson} from '../runJest';

const DIR = path.resolve(__dirname, '../snapshot-mock-fs');
const snapshotDir = path.resolve(DIR, '__tests__/__snapshots__');
const snapshotFile = path.resolve(snapshotDir, 'snapshot.test.js.snap');

beforeEach(() => rimraf.sync(snapshotDir));
afterAll(() => rimraf.sync(snapshotDir));

test('store snapshot even if fs is mocked', () => {
  const {json, status, stderr} = runJestJson(DIR, ['--ci=false']);

  expect(status).toBe(0);
  expect(json.numTotalTests).toBe(1);
  expect(json.numPassedTests).toBe(1);

  expect(stderr).toMatch('1 snapshot written from 1 test suite.');
  expect(wrap(extractSummary(stderr).summary)).toMatchSnapshot();

  // $FlowFixMe dynamic require
  const content = require(snapshotFile);
  expect(content['snapshot 1']).toBe(`
Object {
  "foo": "bar",
}
`);
});
