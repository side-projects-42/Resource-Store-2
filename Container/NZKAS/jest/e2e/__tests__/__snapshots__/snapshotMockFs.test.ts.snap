// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store snapshot even if fs is mocked 1`] = `
Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   1 written, 1 total
Time:        <<REPLACED>>
Ran all test suites.
`;
